Client/Helpers/TabEventHandler.swift

```diff
@@ -0,0 +1,171 @@
+// 3. a TabEventLabel, which is needed for registration
+// 4. a TabEvent, with whatever parameters are needed.
+//    i) a case to map the event to the event label (var label)
+//   ii) a case to map the event to the event handler (func handle:with:)
+protocol TabEventHandler {
+    func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata)
+    func tab(_ tab: Tab, didLoadFavicon favicon: Favicon)
+    func tabDidGainFocus(_ tab: Tab)
+    func tabDidLoseFocus(_ tab: Tab)
+    func tabDidClose(_ tab: Tab)
+}
+
+// Provide default implmentations, because we don't want to litter the code with
+// empty methods, and `@objc optional` doesn't really work very well.
+extension TabEventHandler {
+    func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata) {}
+    func tab(_ tab: Tab, didLoadFavicon favicon: Favicon) {}
+    func tabDidGainFocus(_ tab: Tab) {}
+    func tabDidLoseFocus(_ tab: Tab) {}
+    func tabDidClose(_ tab: Tab) {}
+}
+
+enum TabEventLabel: String {
+    case didLoadPageMetadata = "didLoadPageMetadata"
+    case didLoadFavicon = "didLoadFavicon"
+    case didGainFocus = "didGainFocus"
+    case didLoseFocus = "didLoseFocus"
+    case didClose = "didClose"
+}
+
+enum TabEvent {
+    case didLoadPageMetadata(PageMetadata)
+    case didLoadFavicon(Favicon)
+    case didGainFocus
+    case didLoseFocus
+    case didClose
+
+    var label: TabEventLabel {
+        switch self {
+        case .didLoadPageMetadata:
+            return .didLoadPageMetadata
+        case .didLoadFavicon:
+            return .didLoadFavicon
+        case .didGainFocus:
+            return .didGainFocus
+        case .didLoseFocus:
+            return .didLoseFocus
+        case .didClose:
+            return .didClose
+        }
+    }
+    
+    func handle(_ tab: Tab, with handler: TabEventHandler) {
+        switch self {
+        case .didLoadPageMetadata(let metadata):
+            handler.tab(tab, didLoadPageMetadata: metadata)
+        case .didLoadFavicon(let favicon):
+            handler.tab(tab, didLoadFavicon: favicon)
+        case .didGainFocus:
+            handler.tabDidGainFocus(tab)
+        case .didLoseFocus:
+            handler.tabDidLoseFocus(tab)
+        case .didClose:
+            handler.tabDidClose(tab)
+        }
+    }
+}
+
+// Hide some of the machinery away from the boiler plate above.
+////////////////////////////////////////////////////////////////////////////////////////
+extension TabEventLabel {
+    var name: Notification.Name {
+        return Notification.Name(self.rawValue)
+    }
+}
+
+extension TabEvent {
+    func notification(for tab: Any) -> Notification {
+        return Notification(name: label.name, object: tab, userInfo: ["payload": self])
+    }
+
+    /// Use this method to post notifications to any concerned listeners.
+    static func post(_ event: TabEvent, for tab: Any) {
+        NotificationCenter.default.post(event.notification(for: tab))
```

Comment:
Apple [mentions](https://developer.apple.com/documentation/foundation/notificationcenter/1414169-default) that if our app plans to use the notification center a lot we should make our own channel instead of using `default`